Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,504 @@
+    /// - parameter preference: The `EventLoopPreference` the connection pool will respect to lease a new connection
+    /// - parameter deadline: The connection timeout
+    /// - Returns: A connection  corresponding to the specified parameters
+    ///
+    /// When the pool is asked for a new connection, it creates a `Key` from the url associated to the `request`. This key
+    /// is used to determine if there already exists an associated `ConnectionProvider` in `connectionProviders`
+    /// if it does, the connection provider then takes care of leasing a new connection. If a connection provider doesn't exist, it is created.
+    func getConnection(for request: HTTPClient.Request, preference: HTTPClient.EventLoopPreference, deadline: NIODeadline?) -> EventLoopFuture<Connection> {
+        let key = Key(request: request)
+        // FIXME: Is this correct to use self.loopGroup?
+        switch self.getAction(for: key, eventLoop: self.loopGroup.next()) {
+        case .useExisting(let provider):
+            return provider.getConnection(preference: preference)
+
+        case .make(completing: let providerPromise):
+            // If no connection provider exists for the given key, create a new one
+            let selectedEventLoop = providerPromise.futureResult.eventLoop
+            var bootstrap = ClientBootstrap.makeHTTPClientBootstrapBase(group: selectedEventLoop, host: request.host, port: request.port, configuration: self.configuration) { _ in
+                self.loopGroup.next().makeSucceededFuture(())
+            }
+
+            if let timeout = resolve(timeout: self.configuration.timeout.connect, deadline: deadline) {
+                bootstrap = bootstrap.connectTimeout(timeout)
+            }
+
+            let address = HTTPClient.resolveAddress(host: request.host, port: request.port, proxy: self.configuration.proxy)
+
+            return bootstrap.connect(host: address.host, port: address.port).flatMap { channel in
+                let http1PipelineConfigurator: (ChannelPipeline) -> EventLoopFuture<ConnectionProvider> = { pipeline in
+                    pipeline.addHTTPClientHandlers(leftOverBytesStrategy: .forwardBytes).map { _ in
+                        let http1Provider = ConnectionProvider.http1(HTTP1ConnectionProvider(group: self.loopGroup, key: key, configuration: self.configuration, initialConnection: Connection(key: key, channel: channel, parentPool: self), parentPool: self))
+                        return http1Provider
+                    }
+                }
+
+                return channel.pipeline.addSSLHandlerIfNeeded(for: key, tlsConfiguration: self.configuration.tlsConfiguration).flatMap { _ in
+                    http1PipelineConfigurator(channel.pipeline)
+                }.flatMap { provider in
+                    providerPromise.succeed(provider)
+                    return provider.getConnection(preference: preference)
+                }
```

Comment:
The callback flow from lines `109` to here is a bit weird. Can I propose this layout instead?

```swift
            return bootstrap.connect(host: address.host, port: address.port).flatMap { channel in
                 return channel.pipeline.addSSLHandlerIfNeeded(for: key, tlsConfiguration: self.configuration.tlsConfiguration).flatMap {
                     channel.pipeline.addHTTPClientHandlers(leftOverBytesStrategy: .forwardBytes)
                 }.flatMap {
                     let provider = ConnectionProvider.http1(HTTP1ConnectionProvider(group: self.loopGroup, key: key, configuration: self.configuration, initialConnection: Connection(key: key, channel: channel, parentPool: self), parentPool: self))
                     providerPromise.succeed(provider)
                     return provider.getConnection(preference: preference)
                 }
```

This code is a lot more compact, and more clearly calls out where we can improve performance in future.

Reply:
Thanks, using it in https://github.com/swift-server/async-http-client/pull/105/commits/c8dca4cdf7be7eb88dbfb9c0779d8ecbb368b84d, it was left pretty similar to how it was done when ALPN was still there but it's not relevant anymore.

Code could also be more natural if it was possible to directly put it in the bootstrap `channelInitializer` but AFAICT it cannot be done since channel initialiser may be called multiple times with the current implementation of Happy Eyeballs.